test/docfx.Test/lib/GitHubUserCacheTest.cs

```diff
@@ -0,0 +1,218 @@
+        {
+            VerifyCore(
+                async (cache) => await cache.GetByLogin("alice"),
+                "{\"users\":[{\"id\":1,\"login\":\"alice\",\"name\":\"Alice\",\"emails\":[\"alice@contoso.com\"]}]}",
+                "{\"users\":[{\"id\":1,\"login\":\"alice\",\"name\":\"Alice\",\"emails\":[\"alice@contoso.com\"]}]}",
+                0,
+                0);
+        }
+
+        [Fact]
+        public void GetUserByCommmit()
+        {
+            VerifyCore(
+                async (cache) => await cache.GetByCommit("alice@contoso.com", "owner", "name", "1"),
+                "{}",
+                "{\"users\":[{\"id\":1,\"login\":\"alice\",\"name\":\"Alice\",\"emails\":[\"alice@contoso.com\"]}]}",
+                1,
+                1);
+        }
+
+        [Fact]
+        public void GetSameUserByCommmitMultipleTimesShouldCallGitHubOnce()
+        {
+            VerifyCore(
+                async (cache) =>
+                {
+                    await cache.GetByCommit("alice@contoso.com", "owner", "name", "1");
+                    await cache.GetByCommit("alice@contoso.com", "owner", "name", "2");
+                },
+                "{}",
+                "{\"users\":[{\"id\":1,\"login\":\"alice\",\"name\":\"Alice\",\"emails\":[\"alice@contoso.com\"]}]}",
+                1,
+                1);
+        }
+
+        [Fact]
+        public void GetUserByCommmitFromCache()
+        {
+            VerifyCore(
+                async (cache) => await cache.GetByCommit("alice@contoso.com", "owner", "name", "1"),
+                "{\"users\":[{\"id\":1,\"login\":\"alice\",\"name\":\"Alice\",\"emails\":[\"alice@contoso.com\"]}]}",
+                "{\"users\":[{\"id\":1,\"login\":\"alice\",\"name\":\"Alice\",\"emails\":[\"alice@contoso.com\"]}]}",
+                0,
+                0);
+        }
+
+        [Fact]
+        public void GetUserByCommmitWithNewEmail()
+        {
+            VerifyCore(
+                async (cache) => await cache.GetByCommit("alice_new@contoso.com", "owner", "name", "1"),
+                "{}",
+                "{\"users\":[{\"id\":1,\"login\":\"alice\",\"name\":\"Alice\",\"emails\":[\"alice@contoso.com\",\"alice_new@contoso.com\"]}]}",
+                1,
+                1);
+        }
+
+        [Fact]
+        public void GetUserByCommmitWithNewEmailCanCompleteCache()
+        {
+            VerifyCore(
+                async (cache) => await cache.GetByCommit("alice_new@contoso.com", "owner", "name", "1"),
+                "{\"users\":[{\"id\":1,\"login\":\"alice\",\"name\":\"Alice\",\"emails\":[\"alice@contoso.com\"]}]}",
+                "{\"users\":[{\"id\":1,\"login\":\"alice\",\"name\":\"Alice\",\"emails\":[\"alice@contoso.com\",\"alice_new@contoso.com\"]}]}",
+                0,
+                1);
+        }
+
+        [Fact]
+        public void GetUserByInvalidLogin()
+        {
+            VerifyCore(
+                async (cache) => await cache.GetByLogin("invalid"),
```

Comment:
you can combine `GetUserByInvalidLogin` and `GetUserByInvalidLoginFromCache` by calling `cache.GetByLogin("invalid")` twice and verify there is one outgoing call to GitHub

Reply:
It can ensure invalid user can be well cached on disk between 2 builds if seperates these 2 cases.

---
In reply to: [226194326](https://github.com/dotnet/docfx/pull/3467#discussion_r226194326) [](ancestors = 226194326)